### PR TITLE
chore(ci): make libbpf as the default attacher

### PR DESCRIPTION
**Pull request by vimalk78**
- tag existing 'latest' image as 'latest-bcc'
- tag existing 'latest-libbpf' as 'latest'
- make libbpf image as default released image
- use libbpf as attacher for PR image
- make unit_test run with libbpf attacher
- build kepler with libbpf attacher for rpm build
- rename integration_test to integration_test_bcc
- rename integration_test_libbpf to integration_test

### CODE REVIEW

**Comment by vimalk78** [comment]
Cc: @rootfs @SamYuan1990

File: .github/workflows/image.yml [hunk 1]
```diff
@@ -31,7 +31,7 @@ jobs:
           context: .
           platforms: linux/amd64
           push: true
-          tags: quay.io/sustainable_computing_io/kepler:latest
+          tags: quay.io/sustainable_computing_io/kepler:latest-bcc
           labels: latest
```
Comment (sthaha) [inline .github/workflows/image.yml:35]: latest-bcc ?

File: .github/workflows/image.yml [hunk 2]
```diff
@@ -41,6 +41,6 @@ jobs:
           context: .
           platforms: linux/amd64
           push: true
-          tags: quay.io/sustainable_computing_io/kepler:latest-libbpf
+          tags: quay.io/sustainable_computing_io/kepler:latest
           labels: latest-libbpf
```
Comment (sthaha) [inline .github/workflows/image.yml:45]: latest

File: packaging/rpm/kepler.spec [hunk 3]
```diff
@@ -33,7 +33,7 @@ CROSS_BUILD_BINDIR=_output/bin
 GOARCH=amd64
 %endif
 
-make _build_local GOOS=${GOOS} GOARCH=${GOARCH}
+make _build_local GOOS=${GOOS} GOARCH=${GOARCH} ATTACHER_TAG=libbpf
```
Comment (sthaha) [inline packaging/rpm/kepler.spec:36]: instead, we make the switch in the `Makefile` itself to use `libbpf` as the default attacher?

Comment (vimalk78) [reply packaging/rpm/kepler.spec:36]: perhaps in next PR? lets this change get settled?

File: .github/workflows/image.yml (hunk 1 above)
Comment (vimalk78) [reply .github/workflows/image.yml]: updated

File: .github/workflows/image.yml (hunk 2 above)
Comment (vimalk78) [reply .github/workflows/image.yml]: updated

**Comment by vimalk78** [comment]
@rootfs @marceloamaral can i pls get rights to request a PR review?

File: .github/workflows/integration_test.yml
Comment (vimalk78) [inline .github/workflows/integration_test.yml:1]: this file is is renamed
old name: `integration_test_libbpf.yml`
new name: `integration_test.yml`
there are no other changes in file

File: .github/workflows/integration_test_bcc.yml
Comment (vimalk78) [inline .github/workflows/integration_test_bcc.yml:1]: this file is is renamed
old name: `integration_test.yml`
new name: `integration_test_bcc.yml`
there are no other changes in file

File: .github/workflows/image.yml
Comment (vprashar2929) [inline .github/workflows/image.yml:1]: [This](https://github.com/sustainable-computing-io/kepler/pull/1077/files#diff-113df80dfdb383808d66c98190c3f60ea45745427f926da3b04f572ff081d8a6R28) should be changed to `Build and push kepler latest (bcc)`

Comment (vimalk78) [reply .github/workflows/image.yml:1]: updated

File: packaging/rpm/kepler.spec (hunk 3 above)
Comment (sthaha) [reply packaging/rpm/kepler.spec:36]: works for me :-)

File: .github/workflows/integration_test.yml [hunk 4]
```diff
@@ -24,29 +24,29 @@ jobs:
...
           IMAGE_NAME: "kepler"
+          IMAGE_TAG: "devel"
           CTR_CMD: docker
           IMAGE_OUTPUT_PATH: ${{env.OUTPUT_DIR}}${{env.FILE_NAME}}
       # save kepler image
       - name: save Kepler image as artifact
         uses: actions/upload-artifact@v3
         with:
-          name: kepler
+          name: keplerlibbpf
```
Comment (sthaha) [inline .github/workflows/integration_test.yml:43]:
```suggestion
          name: kepler-libbpf
```

File: .github/workflows/integration_test.yml [hunk 5]
```diff
@@ -59,7 +59,7 @@ jobs:
       - name: download Kepler image artifact
         uses: actions/download-artifact@v3
         with:
-          name: kepler
+          name: keplerlibbpf
```
Comment (sthaha) [inline .github/workflows/integration_test.yml:62]:
```suggestion
          name: kepler-libbpf
```

File: .github/workflows/integration_test_bcc.yml [hunk 6]
```diff
@@ -1,15 +1,15 @@
-name: Integration test with libbpf
+name: Integration test with bcc
 
 on:
   workflow_call:
 
 env:
   OUTPUT_DIR: "_output/"
-  FILE_NAME: "kepler_libbpf.tar.gz"
+  FILE_NAME: "kepler.tar.gz"
```
Comment (sthaha) [inline .github/workflows/integration_test_bcc.yml:8]: kepe
```suggestion
  FILE_NAME: "kepler_bcc.tar.gz"
```

Comment (sthaha) [inline .github/workflows/integration_test_bcc.yml:8]:
```suggestion
  FILE_NAME: "kepler_bcc.tar.gz"
```

File: .github/workflows/integration_test.yml (hunk 4 above)
Comment (vimalk78) [reply .github/workflows/integration_test.yml:43]: these files didnt change in the PR , only got renamed

File: .github/workflows/integration_test_bcc.yml [hunk 7]
```diff
@@ -24,30 +24,30 @@ jobs:
...
           IMAGE_TAG: "devel"
+          IMAGE_NAME: "kepler"
           CTR_CMD: docker
           IMAGE_OUTPUT_PATH: ${{env.OUTPUT_DIR}}${{env.FILE_NAME}}
       # save kepler image
       - name: save Kepler image as artifact
         uses: actions/upload-artifact@v3
         with:
-          name: keplerlibbpf
+          name: kepler
```
Comment (sthaha) [inline .github/workflows/integration_test_bcc.yml:43]:
```suggestion
          name: kepler-bcc
```

File: .github/workflows/integration_test.yml [hunk 8]
```diff
@@ -59,7 +59,7 @@ jobs:
...
 
       - name: build manifest
         run: make build-manifest OPTS="CI_DEPLOY"
```
Comment (sthaha) [inline .github/workflows/integration_test.yml:65]: shouldn't we also have the ATTACHER=libbpf set as well?

File: .github/workflows/integration_test.yml [hunk 9]
```diff
@@ -24,29 +24,29 @@ jobs:
       - name: install libbpf
         uses: sustainable-computing-io/kepler-action@v0.0.3
         with:
-          ebpfprovider: bcc
+          ebpfprovider: libbpf
       # build kepler image
       - name: build and export Kepler image
         run: |
           make build_containerized
```
Comment (sthaha) [inline .github/workflows/integration_test.yml:31]: shouldn't we set the ATTACHER=libbpf for all `make` invocations?

Comment (vimalk78) [reply .github/workflows/integration_test.yml:31]: i do not want  to make too many changes at once.

File: .github/workflows/release.yml [hunk 10]
```diff
@@ -130,17 +130,17 @@ jobs:
           context: .
           platforms: linux/amd64
           push: true
-          tags: quay.io/sustainable_computing_io/kepler:latest, quay.io/sustainable_computing_io/kepler:${{ github.event.inputs.release }}
+          tags: quay.io/sustainable_computing_io/kepler:latest, quay.io/sustainable_computing_io/kepler:${{ github.event.inputs.release }}-bcc
```
Comment (sthaha) [inline .github/workflows/release.yml:133]: should we have the `latest` tag here?

Comment (vimalk78) [reply .github/workflows/release.yml:133]: lets take this question in some other PR

File: .github/workflows/integration_test.yml (hunk 4 above)
Comment (sthaha) [reply .github/workflows/integration_test.yml:43]: Oh!! right!

File: .github/workflows/integration_test.yml (hunk 9 above)
Comment (sthaha) [reply .github/workflows/integration_test.yml:31]: let me rephrase ... currently ` bcc` is the default attach method because this PR doesn't modify Makefile. So, for all make invocations that builds kepler or has dependency on `ATTACHER_TAG` (directly or indirectly), shouldn't we explicitly set `ATTACHER_TAG=libbpf` ?

Comment (vimalk78) [reply .github/workflows/integration_test.yml:31]: i just changed the name of the workflow here. if correct images were built, that should remain same.

Comment (sthaha) [reply .github/workflows/integration_test.yml:31]: Resolving this, since the original libbpf test did not require explicit ATTACHER_TAG

**Review by SamYuan1990 — CHANGES_REQUESTED** [review]
add change request for general question.(may not due to code change, but our release process...)
as we are going to replace bcc by libbpf... hence
when are we going to make a switch? start this PR as middle of 0.6 or once we are get ready for 0.7?
after 0.7, are we going to keep bcc support? if not, all bcc related code can be removed.

File: .github/workflows/release.yml (hunk 10 above)
Comment (sthaha) [reply .github/workflows/release.yml:133]: The reason I ask is because I am assuming that the previous release tagged `bcc` as latest and now we want to switch that to libbpf. Would that be the case if we add `latest` here?

Comment (vimalk78) [reply .github/workflows/release.yml:133]: `tags` field for both images has `quay.io/sustainable_computing_io/kepler:latest`

**Comment by vimalk78** [comment]
@rootfs we wanted this PR to merge before we merged `xgboost` PR https://github.com/sustainable-computing-io/kepler/pull/1052 
now can kepler compile without `xgboost`  ?

**Comment by sthaha** [comment]
@rootfs could we please revert #1052 , wait until this PR gets merged so that a release without xgboost can be made?

**Review by rootfs — APPROVED** [review]

**Comment by rootfs** [comment]
Per @SamYuan1990 comment, the bcc code will held in the source and be removed in the next release. 
Following this PR, the libbpf ebpf module should be also built on the CI. It is currently updated manually.

**Comment by SamYuan1990** [comment]
> Per @SamYuan1990 comment, the bcc code will held in the source and be removed in the next release. Following this PR, the libbpf ebpf module should be also built on the CI. It is currently updated manually.

so we will remove bcc code after 0.8 release?

**Comment by rootfs** [comment]
@vimalk78 we need a CI step to [make genlibbpf](https://github.com/sustainable-computing-io/kepler/blob/main/Makefile#L313)

@SamYuan1990 yes, let's remove all the bcc code in 0.8 once we are fully confident on libbpf.

File: .github/workflows/unit_test.yml [hunk 11]
```diff
@@ -27,18 +27,27 @@ jobs:
...
+        ebpfprovider: libbpf
     - name: Prepare environment 
       run: |
           sudo apt-get install -y cpuid
           cd doc/ && sudo ./dev/prepare_dev_env.sh && cd - 
           git config --global --add safe.directory /kepler
     - name: Run 
       run: |
-          make test-verbose
+          sudo apt remove libbpf-dev
```
Comment (SamYuan1990) [inline .github/workflows/unit_test.yml:41]: if `libbpf-dev` is not enough of our libbpf driver, we should make a deb package at https://github.com/sustainable-computing-io/kepler-ci-artifacts and update our kepler-action.

Comment (SamYuan1990) [reply .github/workflows/unit_test.yml:41]: https://github.com/sustainable-computing-io/kepler-action/issues/87 created accordingly.

Comment (SamYuan1990) [reply .github/workflows/unit_test.yml:41]: @rootfs / @vimalk78 / @sthaha could you please help double confirm if we need sustainable-computing-io/kepler-action#87 ?

File: Makefile [hunk 12]
```diff
@@ -277,6 +277,7 @@ test: ginkgo-set tidy-vendor
 
 test-verbose: ginkgo-set tidy-vendor
 	@echo TAGS=$(GO_BUILD_TAGS)
+	@echo GOENV=$(GOENV)
```
Comment (SamYuan1990) [inline Makefile:280]: do we need echo this?

**Comment by SamYuan1990** [comment]
some further comments been added related pr code but not release process.

File: .github/workflows/unit_test.yml (hunk 11 above)
Comment (SamYuan1990) [reply .github/workflows/unit_test.yml:41]: new issue related to this PR as https://github.com/sustainable-computing-io/kepler/issues/1086

Comment (vimalk78) [reply .github/workflows/unit_test.yml:41]: > if `libbpf-dev` is not enough of our libbpf driver, we should make a deb package at 

yes @SamYuan1990, opening a PR for the same. in this PR i wanted to test what works. will update kepler-action, and then we dont needs this setup in CI

Comment (vimalk78) [reply .github/workflows/unit_test.yml:41]: @SamYuan1990 pls check this PR https://github.com/sustainable-computing-io/kepler-action/pull/90 
Cc: @rootfs

File: Makefile (hunk 12 above)
Comment (vimalk78) [reply Makefile:280]: the information echoed in this line is helpful in checking errors we get from `go build`, if any.
